From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <minetuan2@gmail.com>
Date: Thu, 6 Jan 2022 20:56:59 +0700
Subject: [PATCH] Default Commands change


diff --git a/src/main/java/org/bukkit/command/SimpleCommandMap.java b/src/main/java/org/bukkit/command/SimpleCommandMap.java
index 74252236b138969560e6513f24e7ecc6dc4a4127..023f7c9874bd178ee8dc47a1c3bff688fb9b5063 100644
--- a/src/main/java/org/bukkit/command/SimpleCommandMap.java
+++ b/src/main/java/org/bukkit/command/SimpleCommandMap.java
@@ -36,8 +36,12 @@ public class SimpleCommandMap implements CommandMap {
 
     private void setDefaultCommands() {
         register("bukkit", new VersionCommand("version"));
-        register("bukkit", new ReloadCommand("reload"));
-        register("bukkit", new PluginsCommand("plugins"));
+        // Atelier start - Useless commands
+        if (server.isDebugging()) {
+            register("bukkit", new ReloadCommand("reload"));
+            register("bukkit", new PluginsCommand("plugins"));
+        }
+        // Atelier end
         register("bukkit", new co.aikar.timings.TimingsCommand("timings")); // Paper
     }
 
diff --git a/src/main/java/org/bukkit/command/defaults/VersionCommand.java b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
index 57a21495843f3a144cd73473cdc8781d6129b7ca..3352785d2d3ac12c41ebf9829ca341c54feb70b9 100644
--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -41,7 +41,7 @@ public class VersionCommand extends BukkitCommand {
         super(name);
 
         this.description = "Gets the version of this server including any plugins in use";
-        this.usageMessage = "/version [plugin name]";
+        this.usageMessage = "/version"; // Atelier - remove plugin section
         this.setPermission("bukkit.command.version");
         this.setAliases(Arrays.asList("ver", "about"));
     }
@@ -50,6 +50,14 @@ public class VersionCommand extends BukkitCommand {
     public boolean execute(@NotNull CommandSender sender, @NotNull String currentAlias, @NotNull String[] args) {
         if (!testPermission(sender)) return true;
 
+        // Atelier start - Replace
+        sender.sendMessage(
+            "&3-----------------------------------------",
+            "&fThis server is running on &b☃ &fAtelierMC &fsoftware",
+            "&fMade with determination by MangoStudio x Atelier",
+            "&3-----------------------------------------"
+        );
+        /*
         if (args.length == 0) {
             //sender.sendMessage("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")"); // Paper - moved to setVersionMessage
             sendVersion(sender);
@@ -85,6 +93,7 @@ public class VersionCommand extends BukkitCommand {
                 sender.sendMessage("Use /plugins to get a list of plugins.");
             }
         }
+        */
         return true;
     }
 
