From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <minetuan2@gmail.com>
Date: Wed, 15 Dec 2021 10:30:14 +0700
Subject: [PATCH] isDebugging API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 4eab5f0ab87504c544b4fb574704a7b6394addf6..265838571cb4819edaa39b3484c89602accbd170 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2491,7 +2491,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
     }
 
     public boolean isDebugging() {
-        return false;
+        return SharedConstants.IS_RUNNING_IN_IDE; // AtelierMC
     }
 
     @Deprecated
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index df8dc6f1e10ec95a01578bb9033e33b527a4fb48..71dc6cd1a96872649f6ff9bfa25a7846b0b8cb6d 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -877,7 +877,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     // CraftBukkit start
     public boolean isDebugging() {
-        return this.getProperties().debug;
+        return this.getProperties().debug || SharedConstants.IS_RUNNING_IN_IDE;
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index c038ab5a299cd8d83e6d11f2fc58464f67610d20..393f73dd317e78ad9239f8a8d842546a60d31320 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2805,6 +2805,13 @@ public final class CraftServer implements Server {
         return net.minecraft.server.MinecraftServer.currentTick;
     }
 
+    // AtelierMC start
+    @Override
+    public boolean isDebugging() {
+        return net.minecraft.server.MinecraftServer.getServer().isDebugging();
+    }
+    // AtelierMC end
+
     @Override
     public boolean isStopping() {
         return net.minecraft.server.MinecraftServer.getServer().hasStopped();
