From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <minetuan2@gmail.com>
Date: Wed, 15 Dec 2021 10:30:14 +0700
Subject: [PATCH] isDebugging API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index dd171dc1750a1402fea4f260c09e7d7513465cd5..747cd51b031df94f723f72a945d53d1c7193bb84 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2481,7 +2481,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
     }
 
     public boolean isDebugging() {
-        return false;
+        return SharedConstants.IS_RUNNING_IN_IDE; // AtelierMC
     }
 
     @Deprecated
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 8aaccc7fb7cddfe89e1c123b6f7400c56c5d02e5..654ee1f29b106959c9631c42fea975a9b682a9d5 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -873,7 +873,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     // CraftBukkit start
     public boolean isDebugging() {
-        return this.getProperties().debug;
+        return this.getProperties().debug || SharedConstants.IS_RUNNING_IN_IDE;
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 426974113e73ce07704d3e95c3a123f2fc2f7df2..e65972ce72b344c81c7190eb46fc185fa7045018 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2805,6 +2805,13 @@ public final class CraftServer implements Server {
         return net.minecraft.server.MinecraftServer.currentTick;
     }
 
+    // AtelierMC start
+    @Override
+    public boolean isDebugging() {
+        return net.minecraft.server.MinecraftServer.getServer().isDebugging();
+    }
+    // AtelierMC end
+
     @Override
     public boolean isStopping() {
         return net.minecraft.server.MinecraftServer.getServer().hasStopped();
