From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <minetuan2@gmail.com>
Date: Wed, 15 Dec 2021 10:30:14 +0700
Subject: [PATCH] isDebugging API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 72017d86693f3a1afaabb9a6f491aea043228a5d..9a406cb1b558780425632fbf864d2ae27f43f193 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2479,7 +2479,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
     }
 
     public boolean isDebugging() {
-        return false;
+        return SharedConstants.IS_RUNNING_IN_IDE; // AtelierMC
     }
 
     @Deprecated
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index df8dc6f1e10ec95a01578bb9033e33b527a4fb48..71dc6cd1a96872649f6ff9bfa25a7846b0b8cb6d 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -877,7 +877,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     // CraftBukkit start
     public boolean isDebugging() {
-        return this.getProperties().debug;
+        return this.getProperties().debug || SharedConstants.IS_RUNNING_IN_IDE;
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 426974113e73ce07704d3e95c3a123f2fc2f7df2..e65972ce72b344c81c7190eb46fc185fa7045018 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2805,6 +2805,13 @@ public final class CraftServer implements Server {
         return net.minecraft.server.MinecraftServer.currentTick;
     }
 
+    // AtelierMC start
+    @Override
+    public boolean isDebugging() {
+        return net.minecraft.server.MinecraftServer.getServer().isDebugging();
+    }
+    // AtelierMC end
+
     @Override
     public boolean isStopping() {
         return net.minecraft.server.MinecraftServer.getServer().hasStopped();
