From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <minetuan2@gmail.com>
Date: Wed, 12 Jan 2022 07:41:04 +0700
Subject: [PATCH] (Slice) Ignore fire packets if player has Fire Resistance


diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index b300463bf0e3bc29acc226c85d203a0abc91320a..fa4080cb1b5f7c192e9ca9a2bdf1165c614c433a 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -816,7 +816,13 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, i
 
         this.checkOutOfWorld();
         if (!this.level.isClientSide) {
-            this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            // AtelierMC start - Slice: Dont send fire packets if player has Fire Resistance
+            if (this instanceof net.minecraft.world.entity.LivingEntity livingEntity) {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0 && !livingEntity.hasEffect(net.minecraft.world.effect.MobEffects.FIRE_RESISTANCE));
+            } else {
+                this.setSharedFlagOnFire(this.remainingFireTicks > 0);
+            }
+            // AtelierMC end
         }
 
         this.firstTick = false;
