From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <minetuan2@gmail.com>
Date: Fri, 10 Dec 2021 12:24:01 +0700
Subject: [PATCH] Optimised halloween checker


diff --git a/src/main/java/ga/caomile/atelier/HalloweenChecker.java b/src/main/java/ga/caomile/atelier/HalloweenChecker.java
new file mode 100644
index 0000000000000000000000000000000000000000..04d55f06c5b831da25fcb27e3bb39fc46aebff7b
--- /dev/null
+++ b/src/main/java/ga/caomile/atelier/HalloweenChecker.java
@@ -0,0 +1,56 @@
+package ga.caomile.atelier;
+
+import java.time.LocalDate;
+import java.time.temporal.ChronoField;
+import net.minecraft.server.MinecraftServer;
+
+/**
+ + * Entity halloween checker
+ + * <p>
+ + * Checks whether or not it is halloween at a specific rate rather than every time when
+ + * a entity is being spawned.
+ + * <p>
+ + * The rate changes depending on how much TPS the server has. By default, the rate is every
+ + * 2 hours, or every 144k ticks (if the server has _that_ much uptime)
+ + *
+ + * @author MrIvanPlays
+ + */
+public class HalloweenChecker {
+    private static boolean halloweenSeason = false;
+    private static boolean halloweenDay = false;
+
+    private static int delay = (20 * 60 * 60) * 2;
+    private static int lastCheckTick = -delay;
+
+    public static void tick() {
+        if (MinecraftServer.currentTick % 100 == 0) {
+            // update the delay every 100 ticks
+            if (MinecraftServer.TPS >= 20) {
+                delay = (20 * 60 * 60) * 2;
+            }
+            if (MinecraftServer.TPS < 15) {
+                delay = delay + (20 * 60 * 15);
+            }
+            if (MinecraftServer.TPS < 10) {
+                delay = delay + (20 * 60 * 30);
+            }
+        }
+        if (MinecraftServer.currentTick - lastCheckTick > delay) {
+            LocalDate now = LocalDate.now();
+            int day = now.getDayOfMonth();
+            int month = now.get(ChronoField.MONTH_OF_YEAR);
+
+            halloweenDay = (month == 10) && (day == 31);
+            halloweenSeason = ((month == 10) && (day >= 20)) || ((month == 11) && (day <= 3));
+            lastCheckTick = MinecraftServer.currentTick;
+        }
+    }
+
+    public static boolean isHalloweenSeason() {
+        return halloweenSeason;
+    }
+
+    public static boolean isHalloweenDay() {
+        return halloweenDay;
+    }
+}
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 4dc3b536d01515452c9dd562e3af1b4bef414660..26f7945ddecc728bc6383a04eb475c2c584b4439 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1536,6 +1536,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
             this.runAllTasks();
         }
         // Paper end
+        ga.caomile.atelier.HalloweenChecker.tick(); // Atelier
         // Paper start
         long endTime = System.nanoTime();
         long remaining = (TICK_TIME - (endTime - lastTick)) - catchupTime;
diff --git a/src/main/java/net/minecraft/world/entity/ambient/Bat.java b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
index a76b754305036bf9c6387e27b86ec1559b290774..9583b2360db147829e5bf307a140a3566b09af9b 100644
--- a/src/main/java/net/minecraft/world/entity/ambient/Bat.java
+++ b/src/main/java/net/minecraft/world/entity/ambient/Bat.java
@@ -254,11 +254,15 @@ public class Bat extends AmbientCreature {
         }
     }
 
+    // Atelier start - optimised halloween checker
+    /*
     // Pufferfish start - only check for spooky season once an hour
     private static boolean isSpookySeason = false;
     private static final int ONE_HOUR = 20 * 60 * 60;
     private static int lastSpookyCheck = -ONE_HOUR;
+    */
     private static boolean isHalloween() {
+        /*
         if (net.minecraft.server.MinecraftServer.currentTick - lastSpookyCheck > ONE_HOUR) {
         LocalDate localdate = LocalDate.now();
         int i = localdate.get(ChronoField.DAY_OF_MONTH);
@@ -267,9 +271,11 @@ public class Bat extends AmbientCreature {
         isSpookySeason = j == 10 && i >= 20 || j == 11 && i <= 3;
         lastSpookyCheck = net.minecraft.server.MinecraftServer.currentTick;
         }
-        return isSpookySeason;
+        */
+        return ga.caomile.atelier.HalloweenChecker.isHalloweenSeason();
     }
     // Pufferfish end
+    // Atelier end
 
     @Override
     protected float getStandingEyeHeight(Pose pose, EntityDimensions dimensions) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java b/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java
index e23fe546291e670f89447398507d08a0a07efa85..f04bbe31ed10e5f5ad80d777681541ad43da83f6 100644
--- a/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/AbstractSkeleton.java
@@ -155,14 +155,19 @@ public abstract class AbstractSkeleton extends Monster implements RangedAttackMo
         this.reassessWeaponGoal();
         this.setCanPickUpLoot(this.level.paperConfig.skeletonsAlwaysCanPickUpLoot || this.random.nextFloat() < 0.55F * difficulty.getSpecialMultiplier()); // Paper
         if (this.getItemBySlot(EquipmentSlot.HEAD).isEmpty()) {
+            // Atelier start - optimised halloween checker
+            /*
             LocalDate localdate = LocalDate.now();
             int i = localdate.get(ChronoField.DAY_OF_MONTH);
             int j = localdate.get(ChronoField.MONTH_OF_YEAR);
 
             if (j == 10 && i == 31 && this.random.nextFloat() < 0.25F) {
+            */
+            if (ga.caomile.atelier.HalloweenChecker.isHalloweenDay() && this.random.nextFloat() < 0.25F) {
                 this.setItemSlot(EquipmentSlot.HEAD, new ItemStack(this.random.nextFloat() < 0.1F ? Blocks.JACK_O_LANTERN : Blocks.CARVED_PUMPKIN));
                 this.armorDropChances[EquipmentSlot.HEAD.getIndex()] = 0.0F;
             }
+            // Atelier end
         }
 
         return entityData;
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index e72e9b748b3f3e34baddf01366c703efba50c67c..8bec297f6722fb13924be47d93933c7db8892b4b 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -536,14 +536,19 @@ public class Zombie extends Monster {
         }
 
         if (this.getItemBySlot(EquipmentSlot.HEAD).isEmpty()) {
+            // Atelier start - optimised halloween checker
+            /*
             LocalDate localdate = LocalDate.now();
             int i = localdate.get(ChronoField.DAY_OF_MONTH);
             int j = localdate.get(ChronoField.MONTH_OF_YEAR);
 
             if (j == 10 && i == 31 && this.random.nextFloat() < 0.25F) {
+            */
+            if (ga.caomile.atelier.HalloweenChecker.isHalloweenDay() && this.random.nextFloat() < 0.25F) {
                 this.setItemSlot(EquipmentSlot.HEAD, new ItemStack(this.random.nextFloat() < 0.1F ? Blocks.JACK_O_LANTERN : Blocks.CARVED_PUMPKIN));
                 this.armorDropChances[EquipmentSlot.HEAD.getIndex()] = 0.0F;
             }
+            // Atelier end
         }
 
         this.handleAttributes(f);
