From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <38831897+JustMangoT@users.noreply.github.com>
Date: Sun, 23 Jan 2022 13:04:26 +0700
Subject: [PATCH] Move to worlds folder


diff --git a/src/main/java/net/minecraft/server/Main.java b/src/main/java/net/minecraft/server/Main.java
index 8cf835307c36fb048d62af9f417daeeeb71a0cff..e4bed66e23c63f2f14d2ce3fd8606c9c19bd5e1a 100644
--- a/src/main/java/net/minecraft/server/Main.java
+++ b/src/main/java/net/minecraft/server/Main.java
@@ -144,7 +144,7 @@ public class Main {
                 file = (File) optionset.valueOf("universe"); // CraftBukkit
                 userCacheFile = new File(file, "usercache.json");
             } else {
-                file = new File(bukkitConfiguration.getString("settings.world-container", "."));
+                file = new File(bukkitConfiguration.getString("settings.world-container", "worlds")); // AtelierMC
             }
             // Paper end - fix SPIGOT-5824
             YggdrasilAuthenticationService yggdrasilauthenticationservice = new com.destroystokyo.paper.profile.PaperAuthenticationService(Proxy.NO_PROXY); // Paper
