From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: JustMangoT <minetuan2@gmail.com>
Date: Wed, 15 Dec 2021 10:30:14 +0700
Subject: [PATCH] isDebugging API


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 21f741cd8c439a3b815455b4442e575dae51edac..7c06979f5316224529d678419876ada1e6e9ea4f 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2493,7 +2493,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<Runnab
     }
 
     public boolean isDebugging() {
-        return false;
+        return SharedConstants.IS_RUNNING_IN_IDE; // AtelierMC
     }
 
     @Deprecated
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index 9d4e872cf49407ec9968d9ecab4a6707f6b2d539..7e5bf2ceaa257d236fda121c6398bb1ec1184280 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -864,7 +864,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     // CraftBukkit start
     public boolean isDebugging() {
-        return this.getProperties().debug;
+        return this.getProperties().debug || SharedConstants.IS_RUNNING_IN_IDE;
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 8bc6fb6b578792a3f4583ae194cc16179009bbf1..500cec286758eabcd01145c847af3c9fad954fce 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -2804,6 +2804,13 @@ public final class CraftServer implements Server {
         return net.minecraft.server.MinecraftServer.currentTick;
     }
 
+    // AtelierMC start
+    @Override
+    public boolean isDebugging() {
+        return net.minecraft.server.MinecraftServer.getServer().isDebugging();
+    }
+    // AtelierMC end
+
     @Override
     public boolean isStopping() {
         return net.minecraft.server.MinecraftServer.getServer().hasStopped();
